### PR TITLE
feat: Multiple-Signer Block Format

### PR DESCRIPTION
**Pull request by gregorycoppola**
Implement #4 

Non-compiling draft for discussion.

Are these the right fields do add?

### CODE REVIEW

File: src/chainstate/stacks/block.rs [hunk 1]
```diff
@@ -384,6 +393,7 @@ impl StacksBlock {
             &tx_merkle_root,
             state_index_root,
             microblock_pubkey_hash,
+            &MessageSignatureList::empty(),
```
Comment (gregorycoppola) [inline src/chainstate/stacks/block.rs:396]: change this to come from input

File: src/chainstate/stacks/db/blocks.rs [hunk 2]
```diff
@@ -1000,9 +1003,10 @@ impl StacksChainState {
 
     fn has_blocks_with_microblock_pubkh(
         block_conn: &DBConn,
-        pubkey_hash: &Hash160,
+        pubkey_hashes: &Vec<Hash160>,
         minimum_block_height: i64,
     ) -> bool {
+        let pubkey_hash = pubkey_hashes.last().unwrap();
```
Comment (gregorycoppola) [inline src/chainstate/stacks/db/blocks.rs:1009]: Note: What to do in this case?

File: src/chainstate/stacks/db/transactions.rs [hunk 3]
```diff
@@ -644,12 +645,13 @@ impl StacksChainState {
         })?;
 
         if pkh1 != pkh2 {
-            let msg = format!(
-                "Invalid PoisonMicroblock transaction -- signature pubkey hash {} != {}",
-                &pkh1, &pkh2
-            );
+            let msg = "";
```
Comment (gregorycoppola) [inline src/chainstate/stacks/db/transactions.rs:648]: reinstate this

File: src/chainstate/stacks/db/transactions.rs [hunk 4]
```diff
@@ -711,14 +713,15 @@ impl StacksChainState {
...
                 // public key has never been seen before
-                let msg = format!(
-                    "Invalid Stacks transaction: microblock public key hash {} never seen in this fork",
-                    &pubkh
-                );
-                warn!("{}", &msg;
-                      "microblock_pubkey_hash" => %pubkh
-                );
-
+                //                let msg = format!(
```
Comment (gregorycoppola) [inline src/chainstate/stacks/db/transactions.rs:716]: reinstate

File: src/chainstate/stacks/db/transactions.rs [hunk 5]
```diff
@@ -728,9 +731,9 @@ impl StacksChainState {
                     < current_height
                 {
                     let msg = format!("Invalid Stacks transaction: microblock public key hash from height {} has matured relative to current height {}", height, current_height);
-                    warn!("{}", &msg;
-                          "microblock_pubkey_hash" => %pubkh
-                    );
+                    //                    warn!("{}", &msg;
```
Comment (gregorycoppola) [inline src/chainstate/stacks/db/transactions.rs:734]: reinstate

File: src/chainstate/stacks/db/transactions.rs [hunk 6]
```diff
@@ -755,10 +758,10 @@ impl StacksChainState {
             if mblock_header_1.sequence < seq {
                 // this sender reports a point lower in the stream where a fork occurred, and is now
                 // entitled to a commission of the punished miner's coinbase
-                debug!("Sender {} reports a better poison-miroblock record (at {}) for key {} at height {} than {} (at {})", &sender_principal, mblock_header_1.sequence, &pubkh, mblock_pubk_height, &reporter, seq;
-                    "sender" => %sender_principal,
-                    "microblock_pubkey_hash" => %pubkh
-                );
+                //                debug!("Sender {} reports a better poison-miroblock record (at {}) for key {} at height {} than {} (at {})", &sender_principal, mblock_header_1.sequence, &pubkh, mblock_pubk_height, &reporter, seq;
```
Comment (gregorycoppola) [inline src/chainstate/stacks/db/transactions.rs:761]: reinstate

File: src/chainstate/stacks/db/transactions.rs [hunk 7]
```diff
@@ -790,7 +793,7 @@ impl StacksChainState {
         };
 
         let hash_data = BuffData {
-            data: pubkh.as_bytes().to_vec(),
+            data: pubkh.last().unwrap().as_bytes().to_vec(),
```
Comment (gregorycoppola) [inline src/chainstate/stacks/db/transactions.rs:796]: todo: fix this

File: src/chainstate/stacks/miner.rs [hunk 8]
```diff
@@ -626,6 +627,7 @@ impl StacksBlockBuilder {
             total_work,
             proof,
             &pubkh,
+            &MessageSignatureList::empty(),
```
Comment (gregorycoppola) [inline src/chainstate/stacks/miner.rs:630]: pass this as input, don't hardcode as empty

File: src/core/mempool.rs [hunk 9]
```diff
@@ -1447,6 +1449,7 @@ mod tests {
             tx_merkle_root: Sha512Trunc256Sum::empty(),
             state_index_root: TrieHash::from_empty_data(),
             microblock_pubkey_hash: Hash160([0; 20]),
+            miner_signatures: MessageSignatureList::empty(),
```
Comment (gregorycoppola) [inline src/core/mempool.rs:1452]: remove this.. don't hard-code as empty

**Comment by gregorycoppola** [comment]
@kantai can you re-review this?

I'm not sure where the keys actually come from.

That is, in some cases I put `MessageSignatureList::empty()` as a placeholder argument. But, at some sites, this should be replaced with miner signatures, but I'm not sure which sites that gets put in.

**Comment by kantai** [comment]
> But, at some sites, this should be replaced with miner signatures, but I'm not sure which sites that gets put in.

Basically, these structures should have a couple of implementation methods that enable signing. Those will get invoked by the miner during block construction. These methods would look like `set_block_signature()` and `sign_block(sk: PrivateKey) -> MessageSignature`.

One important thing here is that the input to the signature should be the "block hash", but the signature itself cannot be included in the block hash. So the block hash must be computed without the signature, if that makes sense.

**Comment by gregorycoppola** [comment]
But at which sites does the signing happen?

**Pull request closed without merging**